<a href="https://colab.research.google.com/github/GH-Rumi/PemMes_7_Helmi/blob/main/JS05/JS05_Tugas2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Tugas Lab 2**

**Import Library**

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

**Load dan Preprocessing Data**

In [2]:
# Load Data
df = pd.read_csv('spam.csv', encoding='latin-1')

# Preprocessing
# Drop 3 kolom terakhir
df = df.drop(df.iloc[:,2:], axis=1)
# Ubah nama kolom agar mudah dibaca
df.columns = ['Labels', 'SMS']
# Encode label
df['Labels'] = df['Labels'].map({'spam': 1, 'ham': 0})

# Pisahkan fitur dan label
X = df['SMS'].values
y = df['Labels'].values

# Split data (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)

**Model dengan CountVectorizer (Stop Words Aktif)**

In [3]:
# Inisiasi CountVectorizer dengan stop_words
cv = CountVectorizer(stop_words='english')

# Fitting dan transform data
X_train_cv = cv.fit_transform(X_train)
X_test_cv = cv.transform(X_test)

# Buat model dan latih
mnb_cv = MultinomialNB()
mnb_cv.fit(X_train_cv, y_train)

# Prediksi dan evaluasi
y_pred_cv = mnb_cv.predict(X_test_cv)
acc_cv = accuracy_score(y_test, y_pred_cv)
print(f'Hasil akurasi CountVectorizer: {acc_cv}')

Hasil akurasi CountVectorizer: 0.9829596412556054


**Model dengan TF-IDF (Stop Words Aktif)**

In [4]:
# Inisiasi TfidfVectorizer dengan stop_words
tfidf = TfidfVectorizer(stop_words='english')

# Fitting dan transform data
X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

# Buat model dan latih
mnb_tfidf = MultinomialNB()
mnb_tfidf.fit(X_train_tfidf, y_train)

# Prediksi dan evaluasi
y_pred_tfidf = mnb_tfidf.predict(X_test_tfidf)
acc_tfidf = accuracy_score(y_test, y_pred_tfidf)
print(f'Hasil akurasi TF-IDF: {acc_tfidf}')

Hasil akurasi TF-IDF: 0.9605381165919282


**Evaluasi dan Kesimpulan**

Perbandingan Akurasi:

Akurasi Model CountVectorizer: ~98.30%

Akurasi Model TF-IDF: ~96.05%

**Kesimpulan:**
Berdasarkan hasil pengujian pada kasus dataset spam_2.csv (atau dataset spam.csv), fitur CountVectorizer memberikan hasil yang lebih baik (lebih tinggi) dibandingkan TF-IDF.

**Mengapa ini bisa terjadi?**
Pada kasus deteksi SMS spam, probabilitas suatu pesan merupakan spam sangat dipengaruhi oleh frekuensi kemunculan kata-kata tertentu yang sangat khas (keywords seperti "Free", "Win", "Txt", "Prize", dll). CountVectorizer secara eksplisit mempertahankan nilai count (frekuensi) absolut dari kata-kata tersebut.